l2geth/rollup/sync_service.go

```diff
@@ -683,100 +716,110 @@ func (s *SyncService) ValidateAndApplySequencerTransaction(tx *types.Transaction
+// indices. It does not induce side effects on state
+type indexGetter func() (*uint64, error)
+
+// isAtTip is a function that will determine if the local chain is at the tip
+// of the remote datasource
+func (s *SyncService) isAtTip(index *uint64, get indexGetter) (bool, error) {
+	latest, err := get()
+	if errors.Is(err, errElementNotFound) {
+		if index == nil {
+			return true, nil
+		}
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	// There are no known enqueue transactions locally or remotely
+	if latest == nil && index == nil {
+		return true, nil
+	}
+	// Only one of the transactions are nil due to the check above so they
+	// cannot be equal
+	if latest == nil || index == nil {
+		return false, nil
+	}
+	// The indices are equal
+	if *latest == *index {
+		return true, nil
+	}
+	// The indices are not equal
+	return false, nil
+}
+
+// syncToTip is a function that can be used to sync to the tip of an ordered
+// list of things. It is used to sync transactions, enqueue elements and batches
+func (s *SyncService) syncToTip(sync syncer, getTip indexGetter) error {
 	s.loopLock.Lock()
 	defer s.loopLock.Unlock()
 
 	for {
-		latest, err := s.client.GetLatestTransaction(backend)
+		index, err := sync()
 		if errors.Is(err, errElementNotFound) {
-			log.Info("No transactions to sync")
 			return nil
 		}
 		if err != nil {
-			return fmt.Errorf("Cannot get latest transaction: %w", err)
-		}
-		latestIndex := latest.GetMeta().Index
-		if latestIndex == nil {
-			return errors.New("Latest index is nil")
-		}
-
-		nextIndex := s.GetNextIndex()
-		if nextIndex-1 == *latestIndex {
-			log.Info("No new transactions to sync", "tip", *latestIndex, "backend", backend.String())
-			return nil
-		}
-		if err := s.syncTransactionRange(nextIndex, *latestIndex, backend); err != nil {
 			return err
 		}
-
-		post, err := s.client.GetLatestTransaction(backend)
+		isAtTip, err := s.isAtTip(index, getTip)
 		if err != nil {
-			return fmt.Errorf("Cannot get latest transaction: %w", err)
-		}
-		postLatestIndex := post.GetMeta().Index
-		if postLatestIndex == nil {
-			return errors.New("Latest index is nil")
+			return err
 		}
-
-		if *postLatestIndex == *latestIndex {
+		if isAtTip {
 			return nil
 		}
 	}
 }
 
-// syncTransactionRange will sync a range of transactions from
-// start to end (inclusive) from a specific Backend
-func (s *SyncService) syncTransactionRange(start, end uint64, backend Backend) error {
-	log.Info("Syncing transactions to tip", "start", start, "end", end)
-	for i := start; i <= end; i++ {
-		tx, err := s.client.GetTransaction(i, backend)
-		if err != nil {
-			return fmt.Errorf("cannot fetch transaction %d: %w", i, err)
-		}
-		if err = s.applyTransaction(tx); err != nil {
-			return fmt.Errorf("Cannot apply transaction: %w", err)
-		}
+// sync will sync a range of items
+func (s *SyncService) sync(getLatest indexGetter, getNext nextGetter, sync rangeSyncer) (*uint64, error) {
```

Comment:
Using `sync` as the name of the function and as a function parameter was a bit confusing to me. Maybe we could rename `sync` => `syncer`?

Reply:
Ok, can do